tools/smithy-rs-sync/src/main.rs

```diff
@@ -0,0 +1,409 @@
+
+    file.write(oid.to_string().as_bytes())
+        .with_context(|| format!("Couldn't write commit hash to '{}'", path.to_string_lossy()))?;
+
+    Ok(())
+}
+
+/// Run the necessary commands to build the SDK. On success, returns the path to the folder containing
+/// the build artifacts.
+fn build_sdk(smithy_rs_path: &Path) -> Result<PathBuf> {
+    println!("\tbuilding the SDK...");
+    let start = Instant::now();
+    let gradlew = smithy_rs_path.join("gradlew");
+
+    // The output of running this command isn't logged anywhere unless it fails
+    let clean_command_output = Command::new(&gradlew)
+        .arg("-Paws.fullsdk=true")
+        .arg(":aws:sdk:clean")
+        .current_dir(smithy_rs_path)
+        .output()
+        .with_context(|| {
+            format!(
+                "failed to execute '{} -Paws.fullsdk=true :aws:sdk:clean'",
+                gradlew.to_string_lossy()
+            )
+        })?;
+
+    if !clean_command_output.status.success() {
+        let stderr = String::from_utf8_lossy(&clean_command_output.stderr);
+        let stdout = String::from_utf8_lossy(&clean_command_output.stdout);
+
+        println!("stdout:\n{}\n", stdout);
+        println!("stderr:\n{}\n", stderr);
+
+        bail!("command to clean previous build artifacts from smithy-rs before assembling the SDK returned an error")
+    }
+
+    let assemble_command_output = Command::new(&gradlew)
+        .arg("-Paws.fullsdk=true")
+        .arg(":aws:sdk:assemble")
+        .current_dir(smithy_rs_path)
+        .output()
+        .with_context(|| {
+            format!(
+                "failed to execute '{} -Paws.fullsdk=true :aws:sdk:assemble'",
+                gradlew.to_string_lossy()
+            )
+        })?;
+
+    if !assemble_command_output.status.success() {
+        let stderr = String::from_utf8_lossy(&clean_command_output.stderr);
+        let stdout = String::from_utf8_lossy(&clean_command_output.stdout);
+
+        println!("stdout:\n{}\n", stdout);
+        println!("stderr:\n{}\n", stderr);
+
+        bail!("command to assemble the SDK returned an error")
+    }
+
+    let build_artifact_path = smithy_rs_path.join("aws/sdk/build/aws-sdk");
+
+    println!("\tsuccessfully built the SDK in {:?}", start.elapsed());
+    Ok(build_artifact_path)
+}
+
+/// Delete any current SDK files in aws-sdk-rust. Run this before copying over new files.
+fn clean_out_existing_sdk(aws_sdk_path: &Path) -> Result<()> {
+    println!("\tcleaning out previously built SDK...");
+
+    let sdk_path = format!("{}/sdk/*", aws_sdk_path.to_string_lossy());
+    let remove_sdk_command_output = Command::new("rm")
+        .arg("-rf")
+        .arg(&sdk_path)
+        .current_dir(aws_sdk_path)
+        .output()?;
+    if !remove_sdk_command_output.status.success() {
+        bail!("failed to clean out the SDK folder at {}", sdk_path);
+    }
+
+    let examples_path = format!("{}/example/*", aws_sdk_path.to_string_lossy());
```

Comment:
I always advocate about using `Path` instead of raw strings form paths.

Reply:
The reason I didn't do that here is because of the wildcard. I wasn't sure how this would interact with Rust's `Path` because afaik it's not actually a path but a 'glob'. I'm open to suggestions tho

Reply:
Ah, fair enough, you want globbing here. Disregard my comment.